src/rust/engine/client/src/options/mod.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+mod args;
+mod config;
+mod env;
+mod id;
+mod parse;
+
+use std::collections::{BTreeMap, HashSet};
+use std::ops::Deref;
+use std::os::unix::ffi::OsStrExt;
+use std::path::Path;
+use std::rc::Rc;
+
+use crate::build_root::BuildRoot;
+use crate::option_id;
+
+use self::args::Args;
+use self::config::Config;
+use self::env::Env;
+use self::parse::{parse_bool, parse_string_list};
+
+pub use id::{OptionId, Scope};
+
+#[derive(Copy, Clone, Debug)]
+pub(crate) enum ListEditAction {
+  REPLACE,
+  ADD,
+  REMOVE,
+}
+
+#[derive(Debug)]
+pub(crate) struct ListEdit<T> {
+  pub action: ListEditAction,
+  pub items: Vec<T>,
+}
+
+///
+/// A source of option values.
+///
+/// This is currently a subset of the types of options the Pants python option system handles.
+/// Implementations should mimic the behavior of the equivalent python source.
+///
+trait OptionsSource {
+  ///
+  /// Get a display version of the option `id` that most closely matches the syntax used to supply
+  /// the id at runtime. For example, an global option of "bob" would display as "--bob" for use in
+  /// flag based options and "BOB" in environment variable based options.
+  ///
+  fn display(&self, id: &OptionId) -> String {
+    format!("{}", id)
+  }
+
+  ///
+  /// Get the string option identified by `id` from this source.
+  /// Errors when this source has an option value for `id` but that value is not a string.
+  ///
+  fn get_string(&self, id: &OptionId) -> Result<Option<String>, String>;
+
+  ///
+  /// Get the boolean option identified by `id` from this source.
+  /// Errors when this source has an option value for `id` but that value is not a boolean.
+  ///
+  /// The default implementation looks for a string value for `id` and then attempts to parse it as
+  /// a boolean value.
+  ///
+  fn get_bool(&self, id: &OptionId) -> Result<Option<bool>, String> {
+    if let Some(value) = self.get_string(id)? {
+      parse_bool(&*self.display(id), &*value).map(Some)
+    } else {
+      Ok(None)
+    }
+  }
```

Comment:
Nit: Concrete methods in traits are ok, but it still weirds me out a little bit when they are overridden, because it's less obvious which implementations will be affected by an edit here. This one is overridden in two places.

Reply:
I've minimized this to one trait method that Args and Env truly share.